### ISSUE TITLE
Error applying GAE Firewall Rules 

### ISSUE THREAD

**Issue by goobysnack** [terraform-providers/terraform-provider-google#5681]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
Terraform v0.12.18
"google" (hashicorp/google) 2.20.2

### Affected Resource(s)

* google_app_engine_firewall_rule

### Terraform Configuration Files

```tf
Module call:
module "gcp_gaeapp" {
  source = "git::https://github.<private>/tf-modules/gcp-gaeapp//?ref=master"

  project_id     = module.gcp_project.project_id
  location_id    = "us-central"
  source_range   = ["172.16.100.0/24", "172.16.101.0/24", "172.16.102.0/24", "172.16.103.0/24","172.16.104.0/24"] //not the real IPs  :D 
}

/********************/
main.tf in module gcp-gaeapp:

resource "google_app_engine_application" "app" {
  project       = var.project_id
  location_id   = var.location_id
}

resource "google_app_engine_firewall_rule" "default" {
  project      = google_app_engine_application.app.project
  priority     = 2147483646
  action       = "DENY"
  source_range = "*"
}

// Build FW Rule
resource "google_app_engine_firewall_rule" "rule" {
  count        = length(var.source_range)
  project      = google_app_engine_application.app.project
  action       = "ALLOW"
  priority     = 1000+count.index
  source_range = element(var.source_range, count.index)
}
```
### Expected Behavior

No errors, just creation of the firewall rules.

### Actual Behavior

**Error: Provider produced inconsistent result after apply

When applying changes to
module.gcp_gaeapp.google_app_engine_firewall_rule.rule[0], provider
"registry.terraform.io/-/google" produced an unexpected new value for was
present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.**

The above happens randomly for the different rules.  Rerunning terraform apply eventually resolves the issue.  

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

We use our own credentials, so user auth.

### References

* #0000

**Comment by venkykuberan** [comment]
@goobysnack can you please share the plan output and debug logs to better understand what's being passed to provider and APIs ?

**Comment by goobysnack** [comment]
sorry, that's all the data I have.

**Comment by venkykuberan** [comment]
I am able to repro the issue. I am seeing that API call (POST) to create app_engine_fire_wall_rule is succeeding with 200 response, but the following GET call to pull the fire wall rule is failing. I am attaching the debug log here
[debuglog.txt](https://github.com/terraform-providers/terraform-provider-google/files/4263424/debuglog.txt)

**Comment by TheWolfNL** [comment]
I was running into this issue as well,
Turned on debugging an retrieved a selection of the logs.
I did replace the IP and description 😅 so the content length might not match
```
2020/02/28 13:26:42 [DEBUG] Creating new FirewallRule: map[string]interface {}{"action":"ALLOW", "description":"really good description | test", "priority":4, "sourceRange":"1.2.3.4/32"}
2020/02/28 13:26:42 [DEBUG] Waiting for state to become: [success]
2020/02/28 13:26:42 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
POST /v1/apps/mol-kyc-dev-service-b170/firewall/ingressRules?alt=json HTTP/1.1
Host: appengine.googleapis.com
User-Agent: HashiCorp Terraform/0.12.20 (+https://www.terraform.io) Terraform Plugin SDK/1.4.0 TFC/9836ec2a1f terraform-provider-google/3.10.0
Content-Length: 110
Content-Type: application/json
Accept-Encoding: gzip

{
 "action": "ALLOW",
 "description": "really good description | test",
 "priority": 4,
 "sourceRange": "1.2.3.4/32"
}
2020/02/28 13:26:43 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Fri, 28 Feb 2020 13:26:43 GMT
Etag: c1caf972864e2e0f
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "priority": 4,
er-google_v3.10.0_x4:   "action": "ALLOW",
  "sourceRange": "1.2.3.4/32",
  "description": "really good description | test"
}

2020/02/28 13:26:43 [DEBUG] Finished creating FirewallRule "apps/mol-kyc-dev-service-b170/firewall/ingressRules/4": map[string]interface {}{"action":"ALLOW", "description":"really good description | test", "priority":4, "sourceRange":"1.2.3.4/32"}
2020/02/28 13:26:43 [DEBUG] Waiting for state to become: [success]
2020/02/28 13:26:43 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1/apps/mol-kyc-dev-service-b170/firewall/ingressRules/4?alt=json HTTP/1.1
Host: appengine.googleapis.com
User-Agent: HashiCorp Terraform/0.12.20 (+https://www.terraform.io) Terraform Plugin SDK/1.4.0 TFC/9836ec2a1f terraform-provider-google/3.10.0
Content-Type: application/json
Accept-Encoding: gzip

2020/02/28 13:26:43 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 404 Not Found
Alt-Svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Fri, 28 Feb 2020 13:26:43 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 404,
google_v3.10.0_x4:     "message": "No rule found at position 4.",
    "status": "NOT_FOUND"
  }
}
```

**Comment by TheWolfNL** [comment]
I am also encountering this issue.

**Comment by Karunamon** [comment]
I did a bit of digging on this, and I think it has to do with priorities. Namely, the priority appears to be how Terraform (and the Google API) keeps track of the rule in state.

Using @goobysnack's method of using a count index of a list of ranges to set a priority, I get the same error. More galling, though, is the fact that **random rules within the set are created successfully**. Re-applying the same code over and over and over again will eventually result in all of the rules existing as expected, with a different random set of failures each time.

Another method I tried was `for_each`ing over the list and skipping the priority value outright. This leads to a successful terraform run, but again, with not all rules being created on GCP. As a bonus, even though these rules are in the state file, TF loses track of them and will attempt (and succeed) creating duplicates of the entire list of rules on your next `apply`.

Yet another corner case is that if you don't specify a priority, the GCP API silently takes the highest numbered rule, adds 1000 to it, and uses that as the priority (and doesn't return it to TF, apparently). What this means is that if you've used the workaround mentioned in #3074 for adding a pseudo-default-deny rule (taking the default allow rule's priority of 2147483647 and subtracting one to ensure your new deny rule will always be evaluated last), that App Engine will helpfully take that max number, try to add 1000, fail, and your resources will fail with an error 400.

**Comment by emilymye** [comment]
I'm curious whether this is just a eventual consistency error - I'd be curious to see whether it's really a random issue where the rule doesn't even get added, or if waiting on the rule to exist after creation will solve this. I'll add a PR for that.

**Comment by emilymye** [comment]
(hopefully) fixed by https://github.com/GoogleCloudPlatform/magic-modules/pull/3663 which will most likely get released in 3.28.0
